content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,79 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+- **Reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
+- **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement user solution as an alias or wrapper around reference solution used by tests. Reference solution should not be possible to be called by any other part of the kata other than the tests. In particular, it should not be a global function.  
+Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every requirement in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).
+- Tests should check not only valid inputs, but **invalid inputs and edge conditions should be tested too** (if problem allows them). `null` inputs, empty arrays, empty strings, etc. all should be tested with fixed tests, unless kata description explicitly says that such inputs won't be provided and don't have to be handled by a solution. Random tests are not a replacement for carefully designed assertions for edge cases.
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before corresponding scenario is tested with random tests.
+
+
+## Random Tests
+
+Random tests are not common testing practice in "real life" coding, and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hardcoding, and other workarounds. Goal of random tests is to make input of test cases unpredictable, so only solutions which actually solve the problem can pass.
+
+- **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by a chance.
+- **Randomly generated inputs should be different every time the test suite is run** (with some exceptions, see below). Be sure to seed your RNGs appropriately, and use your random functions correctly, so values used to test user solution are unpredictable. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
```

Comment:
```suggestion
- **Random inputs should be _truly_ (pseudo-)random**, i.e. they should be _dynamically_ generated on every run with the help of suitable library functions such as `Math.random()` in JavaScript or `rand()` in C. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
- In some languages such as C/C++, **make sure that the RNGs are properly seeded with the current time at the beginning** using `srand(time(NULL))` or equivalent; otherwise, the same pseudorandom sequence will be generated on every run and your "random" inputs will devolve into fixed ones.
```

Splitting this into two separate points since "random" inputs as per the XKCD comic and "random" inputs due to forgetting to seed the RNGs are two different types of mistake.

Reply:
Actually I did not consider the XKCD scenario here, because I think I've never seen it in a kata. Do you think it's really necessary?

My goal here was to say "know how to use your RNGs and random APIs", without being too language specific. Notes on `srand` and `time` I wanted to put in separate howto, something like "Writing random tests for [some language]".

Additionally, I am afraid that if we put "truly" here, people might overkill the problem with crypto rngs or external entropy sources like `/dev/urandom/`.

Reply:
shouldn't a specific thing like that be relevant only to snipets provided for that specific language, rather?